<!DOCTYPE html>
<!--[if IE]><![endif]-->
<html>
  
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
    <title>Error Handling &amp; Compensation </title>
    <!-- Prevent search engine web crawlers from indexing or crawling this page -->
    <meta name="robots" content="none">
    <meta name="viewport" content="width=device-width">
    <meta name="title" content="Error Handling &amp; Compensation ">
    <meta name="generator" content="docfx 2.19.1.0">
    
    <link rel="shortcut icon" href="../../images/favicon.ico">
    <link rel="stylesheet" href="../../styles/docfx.vendor.css">
    <link rel="stylesheet" href="../../styles/docfx.css">
    <link rel="stylesheet" href="../../styles/main.css">
    <meta property="docfx:navrel" content="../../toc.html">
    <meta property="docfx:tocrel" content="../toc.html">
    
    
  </head>
  <body data-spy="scroll" data-target="#affix">
    <div id="wrapper">
      <header>
        
        <nav id="autocollapse" class="navbar navbar-inverse ng-scope" role="navigation">
          <div class="container">
            <div class="navbar-header">
              <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbar">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
              </button>
              
              <a class="navbar-brand" href="../../index.html">
                <img id="logo" class="svg" src="../../logo.svg" alt="">
              </a>
            </div>
            <div class="collapse navbar-collapse" id="navbar">
              <form class="navbar-form navbar-right" role="search" id="search">
                <div class="form-group">
                  <input type="text" class="form-control" id="search-query" placeholder="Search" autocomplete="off">
                </div>
              </form>
            </div>
          </div>
        </nav>
        
        <div class="subnav navbar navbar-default">
          <div class="container hide-when-search" id="breadcrumb">
            <ul class="breadcrumb">
              <li></li>
            </ul>
          </div>
        </div>
      </header>
      <div role="main" class="container body-content hide-when-search">
        
        <div class="sidenav hide-when-search">
          <a class="btn toc-toggle collapse" data-toggle="collapse" href="#sidetoggle" aria-expanded="false" aria-controls="sidetoggle">Show / Hide Table of Contents</a>
          <div class="sidetoggle collapse" id="sidetoggle">
            <div id="sidetoc"></div>
          </div>
        </div>
        <div class="article row grid-right">
          <div class="col-md-10">
            <article class="content wrap" id="_content" data-uid="">
<h1 id="error-handling--compensation">Error Handling &amp; Compensation</h1>

<p>Durable Function orchestrations are implemented in pure code and therefore leverage the error handling capabilities of the orchestrator function&#39;s programming language. With this in mind, there really aren&#39;t any new concepts you need to learn about when incorporating error handling and compensation into your orchestrations. However, there are still a few important behaviors to be aware of when planning for error handling.</p>
<h2 id="errors-in-activity-functions">Errors in activity functions</h2>
<p>Any exception that is thrown in an activity function is marshalled back to the orchestrator function and thrown as a <code>TaskFailedException</code>. Users can write the appropriate error handling and compensation code that suits their needs around this.</p>
<p>For example, consider the following orchestrator function which transfers funds from one account to another:</p>
<pre><code class="lang-csharp">#r &quot;Microsoft.Azure.WebJobs.Extensions.DurableTask&quot;

public static async Task Run(DurableOrchestrationContext context)
{
    var transferDetails = ctx.GetInput&lt;TransferOperation&gt;();

    await context.CallFunctionAsync(&quot;DebitAccount&quot;,
        new
        { 
            Account = transferDetails.SourceAccount,
            Amount = transferDetails.Amount
        });

    bool credited;
    try
    {
        await context.CallFunctionAsync(&quot;CreditAccount&quot;,         
            new
            { 
                Account = transferDetails.DestinationAccount,
                Amount = transferDetails.Amount
            });

        credited = true;
    }
    catch (TaskFailedException)
    {
        credited = false;
    }

    if (!credited)
    {
        // Refund the source account.
        // Another try/catch could be used here based on the needs of the application.
        await context.CallFunctionAsync(&quot;CreditAccount&quot;,         
            new
            { 
                Account = transferDetails.SourceAccount,
                Amount = transferDetails.Amount
            });
    }
}
</code></pre><div class="NOTE"><h5>Note</h5><p>The above code requires a DLL reference to DurableTask.Framework assembly, which is not ideal.
<a href="https://github.com/Azure/azure-webjobs-sdk-script-pr/issues/41">https://github.com/Azure/azure-webjobs-sdk-script-pr/issues/41</a></p>
</div>
<p>If the call to the <strong>CreditAccount</strong> function fails for the destination account, the orchestrator function compensates for this by crediting the funds back to the source account.</p>
<h2 id="retry-on-failure">Retry on failure</h2>
<p>There is currently no first-class support for retrying function calls which fail with an error. However, it is still possible to implement retry manually using code, like in the following example:</p>
<pre><code class="lang-csharp">public static async Task&lt;bool&gt; Run(DurableOrchestrationContext context)
{
    const int MaxRetries = 3;

    for (int i = 0; i &lt;= MaxRetries; i++)
    {
        try
        {
            await context.CallFunctionAsync(&quot;FlakyFunction&quot;);
            return true;
        }
        catch { }
    }

    return false;
}
</code></pre><div class="NOTE"><h5>Note</h5><p>Automatic retry is a planned feature for public preview: <a href="https://github.com/Azure/azure-webjobs-sdk-script-pr/issues/42">https://github.com/Azure/azure-webjobs-sdk-script-pr/issues/42</a></p>
</div>
<h2 id="function-timeouts">Function timeouts</h2>
<p>It&#39;s possible that you may want to abandon a function call within an orchestrator function if it is taking too long to complete. The proper way to do this today is by creating a durable timer using <code>context.CreateTimer</code> in conjunction with <code>Task.WhenAny</code>, as in the following example:</p>
<pre><code class="lang-csharp">public static async Task&lt;bool&gt; Run(DurableOrchestrationContext context)
{
    TimeSpan timeout = TimeSpan.FromSeconds(30);
    DateTime deadline = context.CurrentUtcDateTime.Add(timeout);

    using (var cts = new CancellationTokenSource())
    {
        Task activityTask = context.CallFunctionAsync(&quot;FlakyFunction&quot;);
        Task timeoutTask = context.CreateTimer(deadline, cts.Token);

        Task winner = await Task.WhenAny(activityTask, timeoutTask);
        if (winner == activityTask)
        {
            // success case
            cts.Cancel();
            return true;
        }
        else
        {
            // timeout case
            return false;
        }
    }
}
</code></pre><p>For more information on timers, see the <a href="timers.html">Durable Timers</a> topic.</p>
<h2 id="unhandled-exceptions">Unhandled exceptions</h2>
<p>If an orchestrator function fails with an unhandled exception, the details of the exception will be logged and the instance will complete with a <code>Failed</code> status.</p>
<div class="NOTE"><h5>Note</h5><p>Failed instances cannot be recovered currently. However, it would be extremely useful if they could be recovered in certain situations to help mitigate potential data loss.
<a href="https://github.com/Azure/azure-webjobs-sdk-script-pr/issues/20">https://github.com/Azure/azure-webjobs-sdk-script-pr/issues/20</a></p>
</div>
</article>
          </div>
          
          <div class="hidden-sm col-md-2" role="complementary">
            <div class="sideaffix">
              <div class="contribution">
                <ul class="nav">
                  <li>
                    <a href="https://github.com/Azure/azure-functions-durable-extension/blob/master/docfx/articles/topics/error-handling.md/#L1" class="contribution-link">Improve this Doc</a>
                  </li>
                </ul>
              </div>
              <nav class="bs-docs-sidebar hidden-print hidden-xs hidden-sm affix" id="affix">
              <!-- <p><a class="back-to-top" href="#top">Back to top</a><p> -->
              </nav>
            </div>
          </div>
        </div>
      </div>
      
      <footer>
        <div class="grad-bottom"></div>
        <div class="footer">
          <div class="container">
            <span class="pull-right">
              <a href="#top">Back to top</a>
            </span>
            Copyright © Microsoft Corporation
            
          </div>
        </div>
      </footer>
    </div>
    
    <script type="text/javascript" src="../../styles/docfx.vendor.js"></script>
    <script type="text/javascript" src="../../styles/docfx.js"></script>
    <script type="text/javascript" src="../../styles/main.js"></script>
  </body>
</html>
